app/controllers/versions_controller.rb

```diff
@@ -14,4 +15,12 @@ def show
     @on_version_page = true
     render "rubygems/show"
   end
+
+  private
+
+  def set_oldest_version_date
+    oldest_created_at = @rubygem.versions.order(:created_at).first
+    oldest_built_at = @rubygem.versions.order(:built_at).first
+    @oldest_version_date = [oldest_created_at, oldest_built_at].compact.map(&:authored_at).min
```

Comment:
Would it be possible to move this logic into SQL side using MIN function and get the oldest version date in 1 query directly? :thinking:

Reply:
I also checked the performance of these 2 queries vs the min+least or case statement. Surprisingly, at least on my local database, while the query planner thinks the separate checks on `created_at` and `built_at` have lower cost, the execution time of either the min+least or case statement are both much shorter.

```sql
rubygems_development=# explain analyze select min(built_at) from versions where rubygem_id=183850;
                                                                            QUERY PLAN                                                            >
-------------------------------------------------------------------------------------------------------------------------------------------------->
 Result  (cost=41.96..41.97 rows=1 width=8) (actual time=2365.886..2365.896 rows=1 loops=1)
   InitPlan 1 (returns $0)
     ->  Limit  (cost=0.43..41.96 rows=1 width=8) (actual time=2365.874..2365.875 rows=1 loops=1)
           ->  Index Scan using index_versions_on_built_at on versions  (cost=0.43..429251.91 rows=10336 width=8) (actual time=2365.871..2365.872 >
                 Index Cond: (built_at IS NOT NULL)
                 Filter: (rubygem_id = 183850)
                 Rows Removed by Filter: 1071279
 Planning Time: 1.692 ms
 Execution Time: 2366.057 ms
(9 rows)
```

```sql
rubygems_development=# explain analyze select min(created_at) from versions where rubygem_id=183850;

                                                                             QUERY PLAN                                                           >
-------------------------------------------------------------------------------------------------------------------------------------------------->
 Result  (cost=41.88..41.89 rows=1 width=8) (actual time=2718.618..2718.619 rows=1 loops=1)
   InitPlan 1 (returns $0)
     ->  Limit  (cost=0.43..41.88 rows=1 width=8) (actual time=2718.605..2718.606 rows=1 loops=1)
           ->  Index Scan using index_versions_on_created_at on versions  (cost=0.43..428426.01 rows=10336 width=8) (actual time=2718.604..2718.60>
                 Index Cond: (created_at IS NOT NULL)
                 Filter: (rubygem_id = 183850)
                 Rows Removed by Filter: 1068171
 Planning Time: 22.439 ms
 Execution Time: 2719.168 ms
(9 rows)
```


```sql
rubygems_development=# explain analyze select min(least(created_at, built_at)) from versions where rubygem_id=183850;
                                                                      QUERY PLAN                                                                  >
-------------------------------------------------------------------------------------------------------------------------------------------------->
 Aggregate  (cost=30761.62..30761.63 rows=1 width=8) (actual time=187.097..187.098 rows=1 loops=1)
   ->  Bitmap Heap Scan on versions  (cost=196.53..30709.94 rows=10336 width=16) (actual time=20.778..185.133 rows=10991 loops=1)
         Recheck Cond: (rubygem_id = 183850)
         Heap Blocks: exact=6022
         ->  Bitmap Index Scan on index_versions_on_rubygem_id  (cost=0.00..193.95 rows=10336 width=0) (actual time=18.283..18.284 rows=10991 loop>
               Index Cond: (rubygem_id = 183850)
 Planning Time: 0.541 ms
 Execution Time: 187.278 ms
(8 rows)
```


```sql
rubygems_development=# explain analyze select min(case when created_at::date = '2009-07-25'::date and built_at < created_at then built_at else created_at end) as authored_at from versions where rubygem_id=183850;
                                                                     QUERY PLAN                                                                   >
-------------------------------------------------------------------------------------------------------------------------------------------------->
 Aggregate  (cost=30813.30..30813.31 rows=1 width=8) (actual time=94.130..94.132 rows=1 loops=1)
   ->  Bitmap Heap Scan on versions  (cost=196.53..30709.94 rows=10336 width=16) (actual time=4.185..90.765 rows=10991 loops=1)
         Recheck Cond: (rubygem_id = 183850)
         Heap Blocks: exact=6022
         ->  Bitmap Index Scan on index_versions_on_rubygem_id  (cost=0.00..193.95 rows=10336 width=0) (actual time=3.377..3.378 rows=10991 loops=>
               Index Cond: (rubygem_id = 183850)
 Planning Time: 2.007 ms
 Execution Time: 94.376 ms
(8 rows)
```

Reply:
@albertchae thanks for detailed research on this. I'm also using `sorbet-static` for "extreme" cases performance testing, good choice!

I did quick check in production DB and I see similar pattern. It is most likely caused by uneven distribution of versions (thanks for this extreme `sorbet-static` gem). I did quick test also on more common case like `nokogiri` gem and `select min(case when created_at::date = '2009-07-25'::date and built_at < created_at then built_at else created_at 
end) as authored_at from versions where rubygem_id=?` query works well in all cases and if I understand it well, it will keep the same value logic as today (but with bug fixed).

So for now, I would prefer to add `@oldest_version_date = @rubygem.versions.minimum("case when DATE(created_at) = '#{Version::RUBYGEMS_IMPORT_DATE}' and built_at < created_at then built_at else created_at end")` (maybe hide the SQL into model somehow) in controller.

And for the future, we can check if there's any way to synchronise dates in DB with `RUBYGEMS_IMPORT_DATE` constant and normalize them in DB so one `MIN(...)` clause will be enough.

WDYT?

Reply:
> So for now, I would prefer to add `@oldest_version_date = @rubygem.versions.minimum("case when DATE(created_at) = '#{Version::RUBYGEMS_IMPORT_DATE}' and built_at < created_at then built_at else created_at end")` (maybe hide the SQL into model somehow) in controller.

Sounds good. I pushed up a commit to add this as a scope to the `Version` model. One small note, I changed the query to use `<=` to match the application code exactly, but I think we get the same result with or without the `=`
 
> And for the future, we can check if there's any way to synchronise dates in DB with `RUBYGEMS_IMPORT_DATE` constant and normalize them in DB so one `MIN(...)` clause will be enough.

This sounds like a good long term fix!

Reply:
BTW I pushed it up as a separate commit for ease of review, but let me know if you'd like me to squash them. I also edited the PR description to match the new implementatino.